tools/lib.py

```diff
@@ -0,0 +1,36 @@
+import sys
+from threading import Lock
+from typing import Any, Callable
+
+# Unbounded function cache.
+# lru_cache(maxsize=None) has a fast path since it doesn't care about bounds.
+# functools.cache is introduced in python 3.9, but is literally just what the shim is.
+if sys.version_info >= (3, 9):
+    from functools import cache
+else:
+    from functools import lru_cache as _lru_cache
+
+    def cache(func: Callable[..., Any]) -> Any:
+        return _lru_cache(maxsize=None)(func)
+
+
+@cache
+def gitroot() -> str:
+    from os import getcwd
+    from os.path import isdir, normpath
+
+    gitroot, root = getcwd(), "/"
+    while not isdir(f"{gitroot}/.git"):
+        gitroot = normpath(f"{gitroot}/..")
+        if gitroot == root:
+            raise RuntimeError("failed to locate a git root directory")
+    return gitroot
```

Comment:
here's a battle-tested version of this function: https://github.com/pre-commit/pre-commit/blob/fb0ccf3546a9cb34ec3692e403270feb6d6033a2/pre_commit/git.py#L51-L73

notably: `.git` isn't always a directory (sometimes it is a file (submodules / subtrees / worktrees))

Reply:
That's cool, but I think we can operate on the assumption that sentry repos `.git`s are directories and keep this (faster?) function?

Reply:
we're maybe thinking about using subtrees for splitting frontend -- and worktrees are kinda nice for having a separate checkout without all the overhead 🤷